File: LeetCode-Python/UniquePathsII.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        if obstacleGrid[0][0] == 1:
            return 0
        obstacleGrid[0][0] = 1
        
        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    continue
                if obstacleGrid[i][j] == 1:
                    obstacleGrid[i][j] = 0
                else:
                    top = obstacleGrid[i-1][j] if i > 0 else 0
                    left = obstacleGrid[i][j-1] if j > 0 else 0
                    obstacleGrid[i][j] = top + left
        
        return obstacleGrid[-1][-1]

File: LeetCode-Cpp/UniquePathsII.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
         int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        
        // Eğer başlangıç veya bitiş noktası engelli ise, hiç yol yoktur
        if (obstacleGrid[0][0] == 1 || obstacleGrid[m-1][n-1] == 1)
            return 0;
        
        vector<vector<int>> dp(m, vector<int>(n, 0));
        dp[0][0] = 1; // Başlangıç noktası
        
        // İlk sütunu doldur
        for (int i = 1; i < m; ++i)
            dp[i][0] = (obstacleGrid[i][0] == 0 && dp[i-1][0] == 1) ? 1 : 0;
        
        // İlk satırı doldur
        for (int j = 1; j < n; ++j)
            dp[0][j] = (obstacleGrid[0][j] == 0 && dp[0][j-1] == 1) ? 1 : 0;
        
        // Diğer hücreleri doldur
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (obstacleGrid[i][j] == 0)
                    dp[i][j] = dp[i-1][j] + dp[i][j-1];
            }
        }
        
        return dp[m-1][n-1];

    }
};
